esmtools/stats.py

```diff
@@ -135,317 +132,381 @@ def area_weight(da, area_coord="area"):
+        xarray object: Slope, intercept, correlation, p value, and standard error for
+            the linear regression. These 5 parameters are added as a new dimension
+            "parameter".
 
-    References (for dealing with autocorrelation):
-    ----------
-    1. Wilks, Daniel S. Statistical methods in the atmospheric sciences.
-    Vol. 100. Academic press, 2011.
-    2. Lovenduski, Nicole S., and Nicolas Gruber. "Impact of the Southern
-    Annular Mode on Southern Ocean circulation and biology." Geophysical
-    Research Letters 32.11 (2005).
-    3. Brady, R. X., Lovenduski, N. S., Alexander, M. A., Jacox, M., and
-    Gruber, N.: On the role of climate modes in modulating the air-sea CO2
-    fluxes in Eastern Boundary Upwelling Systems, Biogeosciences Discuss.,
-    https://doi.org/10.5194/bg-2018-415, in review, 2018.
     """
-    # Broadcasts a time series to the same coordinates/size as the grid. If they
-    # are both grids, this function does nothing and isn't expensive.
-    x, y = xr.broadcast(x, y)
+    _check_y_not_independent_variable(y, dim)
+    x, slope_factor = _convert_time_and_return_slope_factor(x, dim)
+
+    def _linregress(x, y, slope_factor, nan_policy):
+        x, y = _handle_nans(x, y, nan_policy)
+        m, b, r, p, e = scipy.stats.linregress(x, y)
+        # Multiply slope by factor. If time indices converted to numeric units, this
+        # gets them back to the original units.
+        m *= slope_factor
+        e *= slope_factor
+        return np.array([m, b, r, p, e])
 
-    # Negative lead should have y lead x.
-    if lead < 0:
-        lead = np.abs(lead)
-        return st.corr(y, x, dim=dim, lag=lead, return_p=return_p)
-    else:
-        return st.corr(x, y, dim=dim, lag=lead, return_p=return_p)
+    results = xr.apply_ufunc(
+        _linregress,
+        x,
+        y,
+        slope_factor,
+        nan_policy,
+        vectorize=True,
+        dask="parallelized",
+        input_core_dims=[[dim], [dim], [], []],
+        output_core_dims=[["parameter"]],
+        output_dtypes=["float64"],
+        output_sizes={"parameter": 5},
+    )
+    results = results.assign_coords(
+        parameter=["slope", "intercept", "rvalue", "pvalue", "stderr"]
+    )
+    _warn_if_not_converted_to_original_time_units(x)
+    return results
 
 
 @is_xarray(0)
-def rm_poly(da, order, dim="time"):
-    """
-    Returns xarray object with nth-order fit removed from every time series.
-    Input
-    -----
-    da : xarray DataArray
-        Single time series or many gridded time series of object to be
-        detrended
-    order : int
-        Order of polynomial fit to be removed. If 1, this is functionally
-        the same as calling `rm_trend`
-    dim : str (default 'time')
-        Dimension over which to remove the polynomial fit.
-    Returns
-    -------
-    detrended_ts : xarray DataArray
-        DataArray with detrended time series.
+def polyfit(x, y, order, dim="time", nan_policy="none"):
```

Comment:
I don't think it's any different. I just adapted it since I already had it in `esmtools`. The only difference then is `xr.polyfit` isn't out until the next release.

Reply:
Maybe in the future you can replace. But keep for now

Reply:
polyfit and infer_freq are in the same release

Reply:
Yes I'm trying to figure out if I should replicate `infer_freq` for now because:

1. It doesn't seem easy to require a master branch in `requirements.txt`
2. It will break `climpred` requirements for `esmtools` to want a master branch.

The `infer_freq` code is brief so I'll probably replicate it.